AK/QuickSelect.h

```diff
@@ -0,0 +1,171 @@
+}
+
+}
+
+// This is the Lomuto Partition scheme which is simpler but less efficient than Hoare's partitioning scheme that is traditionally used with quicksort
+// https://en.wikipedia.org/wiki/Quicksort#Lomuto_partition_scheme
+template<typename Collection, typename PivotFn, typename LessThan>
+static size_t partition(Collection& collection, size_t left, size_t right, PivotFn pivot_fn, LessThan less_than)
+{
+    auto pivot_index = pivot_fn(collection, left, right, less_than);
+    auto pivot_value = collection.at(pivot_index);
+    swap(collection.at(pivot_index), collection.at(right));
+    auto store_index = left;
+
+    for (size_t i = left; i < right; i++) {
+        if (less_than(collection.at(i), pivot_value)) {
+            swap(collection.at(store_index), collection.at(i));
+            store_index++;
+        }
+    }
+
+    swap(collection.at(right), collection.at(store_index));
+    return store_index;
+}
+
+template<typename Collection, typename PivotFn, typename LessThan>
+size_t quick_select(Collection& collection, size_t left, size_t right, size_t k, PivotFn pivot_fn, LessThan less_than)
+{
+    // Bail if left is somehow bigger than right and return default constructed result
+    // FIXME: This can also occur when the collection is empty maybe propagate this error somehow?
+    // returning 0 would be a really bad thing since this returns and index and that might lead to memory errors
+    // returning in ErrorOr<size_t> here might be a good option but this is a very specific error that in nearly all circumstances should be considered a bug on the callers site
+    VERIFY(left <= right);
+
+    // If there's only one element, return that element
+    if (left == right)
+        return left;
+
+    auto pivot_index = partition(collection, left, right, pivot_fn, less_than);
+
+    // we found the thing we were searching for
+    if (k == pivot_index)
+        return k;
+
+    // Recurse on the left side
+    if (k < pivot_index)
+        return quick_select(collection, left, pivot_index - 1, k, pivot_fn, less_than);
+
+    // recurse on the right side
+    return quick_select(collection, pivot_index + 1, right, k, pivot_fn, less_than);
+}
+
+//
+template<typename Collection, typename PivotFn, typename LessThan>
+size_t quick_select(Collection& collection, size_t k, PivotFn pivot_fn, LessThan less_than)
+{
+    return quick_select(collection, 0, collection.size() - 1, k, pivot_fn, less_than);
+}
+
+template<typename Collection, typename PivotFn>
+size_t quick_select(Collection& collection, size_t k, PivotFn pivot_fn)
+{
+    return quick_select(collection, 0, collection.size() - 1, k, pivot_fn, [](auto& a, auto& b) { return a < b; });
+}
+
+// All of these quick select implementations return the `index` of the resulting element, after the algorithm has run, not the element itself!
+template<typename Collection>
+size_t quick_select(Collection& collection, size_t k)
```

Comment:
itʼs pretty non-obvious (at least for me) that ‘selectingʼ from the collection modifies it. i feel that at least a comment is in place… (iʼd be the best to reflect the exact side effect in the name in my opinion).

Reply:
Makes sense, i didn't think about it.
I'll add it to the comment.

Reply:
Just as a heads-up, this seems still unresolved, is this right?

Reply:
i thought i did adress it?
Probably got lost somewhere along in the rebase, thanks for reminding me.

Reply:
Now looking at this again since I've already got the repo opened.

At least for me in the diff view of the whole PR, this is showing my new and updated comment already? So I don't know what's going on...

Reply:
Oh! Alright, I thought you wanted to rename it to `quickselect_inplace` or similar, following the tradition to indicate things that might be surprising in the name.

Reply:
Ankles sense and is a good idea. I'll rename all of the variants.

Reply:
Makes not Ankles... Wrote that on my phone with autocorrect...